pep-0592.rst

```diff
@@ -0,0 +1,295 @@
+[tagged-union]_ to Python. One way to accomplish would be to annotate
+``TypedDict`` [typed-dict]_ in Python such that only one field is allowed to be
+set::
+
+    Currency = Annotated[
+        TypedDict('Currency', {'dollars': float, 'pounds': float}, total=False),
+        TaggedUnion,
+    ]
+
+This is a somewhat cumbersome syntax but it allows us to iterate on this
+proof-of-concept and have people with type checkers (or other tools) that don't
+yet support this feature work in a codebase with tagged unions. We could easily
+test this proposal and iron out the kinks before trying to upstream tagged union
+to ``typing``, mypy, etc. Moreover, tools that do not have support for parsing
+the ``TaggedUnion`` annotation would still be able able to treat ``Currency`` as
+a ``TypedDict``, which is still a close approximation (slightly less strict).
+
+Specification
+-------------
+
+Syntax
+~~~~~~
+
+``Annotated`` is parameterized with a type and an arbitrary list of
+Python values that represent the annotations. Here are the specific
+details of the syntax:
+
+* The first argument to ``Annotated`` must be a valid type
+
+* Multiple type annotations are supported (``Annotated`` supports variadic
+  arguments)::
+
+    Annotated[int, ValueRange(3, 10), ctype("char")]
+
+* ``Annotated`` must be called with at least two arguments (
+  ``Annotated[int]`` is not valid)
+
+* The order of the annotations is preserved and matters for equality
+  checks::
+
+    Annotated[int, ValueRange(3, 10), ctype("char")] != Annotated[
+        int, ctype("char"), ValueRange(3, 10)
+    ]
+
+* Nested ``Annotated`` types are flattened, with metadata ordered
+  starting with the innermost annotation::
+
+    Annotated[Annotated[int, ValueRange(3, 10)], ctype("char")] == Annotated[
+        int, ValueRange(3, 10), ctype("char")
+    ]
+
+* Duplicated annotations are not removed::
+
+    Annotated[int, ValueRange(3, 10)] != Annotated[
+        int, ValueRange(3, 10), ValueRange(3, 10)
+    ]
+
+* ``Annotated`` can be used as a higher order aliases::
+
+    Typevar T = ...
+    Vec = Annotated[List[Tuple[T, T]], MaxLen(10)]
+    V = Vec[int]
+
+    V == Annotated[List[Tuple[int, int]], MaxLen(10)]
+
+Consuming annotations
+~~~~~~~~~~~~~~~~~~~~~
+
+Ultimately, the responsibility of how to interpret the annotations (if
+at all) is the responsibility of the tool or library encountering the
+``Annotated`` type. A tool or library encountering an ``Annotated`` type
+can scan through the annotations to determine if they are of interest
+(e.g., using ``isinstance()``).
+
+**Unknown annotations:** When a tool or a library does not support
+annotations or encounters an unknown annotation it should just ignore it
+and treat annotated type as the underlying type. For example, if we were
+to add an annotation that is not an instance of ``new_struct.ctype`` to the
+annotation for name (e.g.,
+``Annotated[str, 'foo', new_struct.ctype("<10s")]``), the unpack method
```

Comment:
Use same name for the hypothetic module here and in the first motivating example.